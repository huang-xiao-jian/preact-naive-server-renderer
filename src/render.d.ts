// package
import { ComponentChild } from 'preact';
// internal
import { HostElement } from './host';

export function render(vnode: ComponentChild, parent: HostElement): HostElement;
