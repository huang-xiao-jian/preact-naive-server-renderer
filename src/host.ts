// internal
import { ElementVNodeProps } from './index.interface';

// inline type
type HostElementChild =
  | HostElement
  | string
  | number
  | boolean
  | null
  | undefined;

// type, props, children
export class HostElement {
  public children: HostElementChild[];

  constructor(public type: string, public props: ElementVNodeProps) {
    this.type = type;
    this.props = props;
    this.children = [];
  }

  appendChild(element: HostElementChild) {
    this.children.push(element);
  }
}

export function createElement(type: string, props: ElementVNodeProps): HostElement {
  return Reflect.construct(HostElement, [type, props]);
}
