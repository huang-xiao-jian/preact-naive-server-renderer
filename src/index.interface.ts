// stringify related
export type SingleTag = string | RegExp;

// spread case
export interface ElementVNode {
  type: string;
  props: ElementVNodeProps & SpecialElementVNodeProps;
  children: ElementChild[];
}

export type ElementChild =
  | ElementVNode
  | string
  | number
  | boolean
  | null
  | undefined;

export interface ElementVNodeProps {
  [key: string]: string | boolean | number;
}

// 渲染成 VNode tree，确定所有 children 都为数组
export interface SpecialElementVNodeProps {
  dangerouslySetInnerHTML?: {
    __html: string;
  };
}

