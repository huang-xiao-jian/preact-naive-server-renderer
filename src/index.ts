/**
 * @description - @coco-platform/init-cli generated template
 * @author - huang.jian <hjj491229492@hotmail.com>
 */

export { render } from './render';
export { stringify } from './stringify';
