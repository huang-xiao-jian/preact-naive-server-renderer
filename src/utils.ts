// internal
import { SingleTag } from './index.interface';

export function isObject(payload: unknown): payload is object {
  return Object.prototype.toString.call(payload) === "[object Object]"
}

// skip children which should not have entity
export function isFalsyChild(child: unknown): boolean {
  return child === undefined || child === null || typeof child === 'boolean' || Number.isNaN(child);
}

export function isSingleTag(tag: string, singleTags: SingleTag[]) {
  return singleTags.some((singleTag) => typeof singleTag === 'string' ? singleTag === tag : singleTag.test(tag))
}
