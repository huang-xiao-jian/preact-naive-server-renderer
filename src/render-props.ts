// internal
import { isObject, styleObjToCss } from './utils';
// interface
import { ElementVNodeProps } from './index.interface';

/**
 * @description - normalize vnode props
 *
 * @param {import(./index.interface).ElementVNodeProps} props - original vnode props
 *
 * @returns {import(./index.interface).ElementVNodeProps}
 */
export function renderProps(props: ElementVNodeProps): ElementVNodeProps {
  const normalizedProps: ElementVNodeProps = {};
  const keys = Reflect.ownKeys(props) as string[];
  // ignore logical property, see https://github.com/preactjs/preact-render-to-string/blob/694155845d7d98637d80f2fd10b92e64412cfe50/src/index.js#L166
  const specialKeys = [
    'defaultValue',
    'key',
    'ref',
    'children',
    '__self',
    '__source',
    'dangerouslySetInnerHTML',
  ];

  /**
   * extra notice:
   *
   * 1. handle boolean attribute within `stringify`；
   * 2. handle html entity within `stringify`；
   * 3. textarea'['value'] --> <textarea value="a&b"> --> <textarea>a&amp;b</textarea> within `stringify`；
   * 4. option['value'] --> <option>value</option> `stringify`；;
   *
   * */
  keys
    // keep order for better test
    .sort()
    /* ignore specific properties */
    // ignore preact related logical properties
    .filter((key) => !specialKeys.includes(key))
    // https://github.com/preactjs/preact-render-to-string/blob/694155845d7d98637d80f2fd10b92e64412cfe50/src/index.js#L164
    .filter((key) => !/[\s\n\\/='"\0<>]/.test(key))
    // ignore function payload and null, undefined
    .filter((key) => props[key] != null && typeof props[key] !== 'function')
    .forEach((key) => {
      const payload = props[key];

      /* key reverse */
      if (key === 'className') {
        normalizedProps.class = payload;
      } else if (key === 'htmlFor') {
        normalizedProps.for = payload;
      }
      /* payload convert */
      // only convert style object
      else if (key === 'style' && isObject(payload)) {
        normalizedProps.style = styleObjToCss(payload);
      }
      // fallback pass through
      else {
        normalizedProps[key] = payload;
      }
    });

  return normalizedProps;
}
