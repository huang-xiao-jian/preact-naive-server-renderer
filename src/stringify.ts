/**
 * @description - stringify element vnode tree into string without property crop logical
 * @author - huang.jian <hjj491229492@hotmail.com>
 */

// internal
import {
  ElementVNode,
  ElementChild,
  SingleTag,
  ElementVNodeProps,
} from './index.interface';
import { isSingleTag, encodeEntities } from './utils';

// scope
export const DEFAULT_SINGLE_TAGS = [
  'area',
  'base',
  'br',
  'col',
  'command',
  'embed',
  'hr',
  'img',
  'input',
  'keygen',
  'link',
  'menuitem',
  'meta',
  'param',
  'source',
  'track',
  'wbr',
];

export function stringifyElementAttributes(props: ElementVNodeProps) {
  // avoid annoying type mismatch issue
  const keys = Reflect.ownKeys(props) as string[];
  const result = keys.sort().reduce<string>((acc, key) => {
    const value = props[key as string];

    if (typeof value === 'string') {
      // quote attributes
      return value === ''
        ? `${acc} ${key}`
        : `${acc} ${key}="${encodeEntities(value)}"`;
    }

    if (typeof value === 'number') {
      return `${acc} ${key}=${value}`;
    }

    if (typeof value === 'boolean') {
      if (value === true) {
        return `${acc} ${key}`;
      }
      // always use string values instead of booleans for aria attributes
      else if (key[0] === 'a' && key[1] === 'r') {
        return `${acc} ${key}=${value}`;
      }
    }

    return acc;
  }, '');

  return result.trim();
}

export function stringifyElementChild(
  child: ElementChild,
  singleTags: SingleTag[] = []
) {
  // number convert into string
  if (typeof child === 'string' || typeof child === 'number') {
    return encodeEntities(`${child}`);
  }

  // stringify element vnode
  const {
    type,
    children,
    props: { dangerouslySetInnerHTML, ...rest },
  } = child as ElementVNode;
  const before = `<${type}`;
  const attributes = stringifyElementAttributes(rest);

  // self close tag, skip children
  if (isSingleTag(type, singleTags)) {
    return `${before} ${attributes} />`;
  }

  return `${before} ${attributes}>${stringify(children)}</${type}>`;
}

export function stringify(
  children: ElementChild[],
  singleTags: SingleTag[] = DEFAULT_SINGLE_TAGS
): string {
  return children
    .map((child) => stringifyElementChild(child, singleTags))
    .join('');
}
