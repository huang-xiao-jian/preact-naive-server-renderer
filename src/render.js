/**
 * @description - render vnode tree into element vnode tree
 * @author - huang.jian <hjj491229492@hotmail.com>
 */

// package
import { toChildArray, Component, options } from 'preact';
// internal
import { createElement } from './host';
import { renderProps } from './render-props';

// server side render entrance
export function render(vnode, parent, globalContext = {}) {
  if (typeof vnode === 'string' || typeof vnode === 'number') {
    renderTextNode(vnode, parent);
  } else if (typeof vnode.type === 'function') {
    renderFunctionalNode(vnode, parent, globalContext);
  } else if (typeof vnode.type === 'string') {
    renderElementNode(vnode, parent, globalContext);
  } else if (Array.isArray(vnode)) {
    renderChildren(vnode, parent, globalContext);
  }
  // else {
  //   ignore boolean, undefined, null
  //}
}

// just for illustrate
export function renderTextNode(vnode, parent) {
  parent.appendChild(`${vnode}`);
}

export function renderElementNode(vnode, parent, globalContext) {
  const { type, props } = vnode;
  // render properties earlier
  const properties = renderProps(props);
  const element = createElement(type, properties);

  // handle special prop case dangerouslySetInnerHTML
  if (props.dangerouslySetInnerHTML?.__html) {
    element.appendChild(props.dangerouslySetInnerHTML.__html);
  } else {
    // render children into element, consider it as standalone render context
    renderChildren(props.children, element, globalContext);
  }

  // insert operation, cursor always tail within server renderer
  parent.appendChild(element);
}

export function renderChildren(children, parent, globalContext) {
  // toChildArray flatten children, ignore null, undefined, boolean child
  // keep only string, number and element vnode
  const _children = toChildArray(children);

  _children.forEach((childVNode) => {
    render(childVNode, parent, globalContext);
  });
}

export function renderFunctionalNode(vnode, parent, globalContext) {
  // 组件实例
  let ins;

  const contextType = vnode.type.contextType;
  const provider = contextType && globalContext[contextType.__c];
  // contextType should only related to class component
  // useContext should only related to function component
  // context structure change within server side, weired
  // _defaultValue --> __
  const context = contextType
    ? provider
      ? provider.props.value
      : contextType.__
    : globalContext;

  // Instantiate the new component
  if (vnode.type.prototype && vnode.type.prototype.render) {
    ins = new vnode.type(vnode.props, context);
  } else {
    ins = new Component(vnode.props, context);
    ins.constructor = vnode.type;
    ins.render = doRender;
  }

  // preset property
  ins.state = ins.state || {};

  // only getDerivedStateFromProps hook should be called
  if (typeof vnode.type.getDerivedStateFromProps === 'function') {
    Object.assign(
      ins.state,
      vnode.type.getDerivedStateFromProps(vnode.props, ins._nextState)
    );
  }

  // hooks bounding property change, weired still
  // unnecessary when drop hooks support
  // __component --> __c
  vnode.__c = ins;

  // context structure change within server side, weired
  // options.render --> options.__r
  if (options.__r) {
    options.__r(vnode);
  }

  const children = ins.render(ins.props, ins.state, ins.context);

  // support context inheritance
  // getChildContext only occure after render method call
  if (ins.getChildContext != null) {
    globalContext = { ...globalContext, ...ins.getChildContext() };
  }

  renderChildren(children, parent, globalContext);
}

/** The `.render()` method for a PFC backing instance. */
function doRender(props) {
  return this.constructor(props);
}
