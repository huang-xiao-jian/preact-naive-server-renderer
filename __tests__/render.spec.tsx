// package
import { h, Fragment } from 'preact';
// internal
import { render } from '../src/render';
import { createElement } from '../src/host';
// fixtures
import { A, B, C, Theme } from './__fixtures__/components.fixture';

// Todo - find better way to keep import references
if (h || Fragment) {
}

describe('renderer', () => {
  const parent = createElement('root', {});

  beforeEach(() => {
    parent.children = [];
  });

  it('should render naive element vnode', () => {
    render(
      <p className="text-left">Left aligned text on all viewport sizes.</p>,
      parent
    );

    expect(parent).toMatchSnapshot();
  });

  it('should render naive functinal vnode', () => {
    render(
      <A title="Hello World" message="youth is not a time of life!" />,
      parent
    );

    expect(parent).toMatchSnapshot();
  });

  it('should support mixed children', () => {
    const vnode = (
      <>
        {null}
        <p className="text-left">Left aligned text on all viewport sizes.</p>
        {undefined}
        {null}
        <A title="Hello World" message="youth is not a time of life!" />
        {true}
        Mixed Text Node
      </>
    );

    render(vnode, parent);

    expect(parent).toMatchSnapshot();
  });

  it('should support context type with defaults', () => {
    render(<B />, parent);

    expect(parent).toMatchSnapshot();
  });

  it('should support context type with payload', () => {
    render(
      <Theme.Provider value={{ header: 'yellow' }}>
        <B />
      </Theme.Provider>,
      parent
    );

    expect(parent).toMatchSnapshot();
  });

  it('should support context within hooks', () => {
    render(
      <Theme.Provider value={{ header: 'yellow' }}>
        <C />
      </Theme.Provider>,
      parent
    );

    expect(parent).toMatchSnapshot();
  });
});
