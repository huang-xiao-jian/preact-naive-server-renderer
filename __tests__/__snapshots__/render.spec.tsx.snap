// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderer should render naive element vnode 1`] = `
HostElement {
  "children": Array [
    HostElement {
      "children": Array [
        "Left aligned text on all viewport sizes.",
      ],
      "props": Object {
        "class": "text-left",
      },
      "type": "p",
    },
  ],
  "props": Object {},
  "type": "root",
}
`;

exports[`renderer should render naive functinal vnode 1`] = `
HostElement {
  "children": Array [
    HostElement {
      "children": Array [
        "Hello World",
      ],
      "props": Object {},
      "type": "h4",
    },
    HostElement {
      "children": Array [
        "youth is not a time of life!",
      ],
      "props": Object {},
      "type": "p",
    },
  ],
  "props": Object {},
  "type": "root",
}
`;

exports[`renderer should support context type with defaults 1`] = `
HostElement {
  "children": Array [
    HostElement {
      "children": Array [
        "Header With Context red",
      ],
      "props": Object {},
      "type": "header",
    },
  ],
  "props": Object {},
  "type": "root",
}
`;

exports[`renderer should support context type with payload 1`] = `
HostElement {
  "children": Array [
    HostElement {
      "children": Array [
        "Header With Context yellow",
      ],
      "props": Object {},
      "type": "header",
    },
  ],
  "props": Object {},
  "type": "root",
}
`;

exports[`renderer should support context within hooks 1`] = `
HostElement {
  "children": Array [
    HostElement {
      "children": Array [
        "Header With Context yellow",
      ],
      "props": Object {},
      "type": "header",
    },
  ],
  "props": Object {},
  "type": "root",
}
`;

exports[`renderer should support mixed children 1`] = `
HostElement {
  "children": Array [
    HostElement {
      "children": Array [
        "Left aligned text on all viewport sizes.",
      ],
      "props": Object {
        "class": "text-left",
      },
      "type": "p",
    },
    HostElement {
      "children": Array [
        "Hello World",
      ],
      "props": Object {},
      "type": "h4",
    },
    HostElement {
      "children": Array [
        "youth is not a time of life!",
      ],
      "props": Object {},
      "type": "p",
    },
    "Mixed Text Node",
  ],
  "props": Object {},
  "type": "root",
}
`;
